src/geoFire/geoQuery.ts

```diff
@@ -0,0 +1,526 @@
+import * as firebase from 'firebase';
+
+import { GeoCallbackRegistration } from './geoCallbackRegistration';
+import { decodeGeoFireObject, distance, encodeGeohash, geoFireGetKey, geohashQueries, validateCriteria, validateLocation } from './geoFireUtils';
+
+import { QueryCriteria } from './interfaces';
+
+/**
+ * Creates a GeoQuery instance.
+ */
+export class GeoQuery {
+  // Event callbacks
+  private _callbacks: any = { ready: [], key_entered: [], key_exited: [], key_moved: [] };
+  // Variable to track when the query is cancelled
+  private _cancelled: boolean = false;
+  private _center: number[];
+  // A dictionary of geohash queries which currently have an active callbacks
+  private _currentGeohashesQueried: any = {};
+  // A dictionary of locations that a currently active in the queries
+  // Note that not all of these are currently within this query
+  private _locationsTracked: any = {};
+  private _radius: number;
+
+  // Variables used to keep track of when to fire the 'ready' event
+  private _valueEventFired: boolean = false;
+  private _outstandingGeohashReadyEvents: any;
+  // Every ten seconds, clean up the geohashes we are currently querying for. We keep these around
+  // for a little while since it's likely that they will need to be re-queried shortly after they
+  // move outside of the query's bounding box.
+  private _geohashCleanupScheduled: boolean = false;
+  private _cleanUpCurrentGeohashesQueriedInterval: NodeJS.Timer;
```

Comment:
Typing the return of a setInterval is scary in that the return type is different for Node and Browser envs. Maybe consider **not** typing this specific thing?